core/src/labels/labels.cpp

```diff
@@ -221,60 +208,121 @@ void Labels::update(const View& _view, float _dt,
 
     /// Collect and update labels from visible tiles
 
-    m_needUpdate = updateLabels(_styles, _tiles, _dt, dz, _view);
+    updateLabels(_styles, _tiles, _dt, dz, _view);
+
+
+    /// Mark labels to skip transitions
+
+    if (int(m_lastZoom) != int(_view.getZoom())) {
```

Comment:
Something like this might work better:

- checkSkipTransition after collision resolution.
 - this will make sure only "non-colliding/visible" labels go to skip-transition state
- in evalState, if a label's state is skipTransition... set its state to "visible".